privacyidea/lib/tokens/tantoken.py

```diff
@@ -0,0 +1,190 @@
+                              'of paper.',
+               'init': {},
+               'config': {},
+               'user':  ['enroll'],
+               # This tokentype is enrollable in the UI for...
+               'ui_enroll': ["admin", "user"],
+               'policy': {
+                   SCOPE.ENROLL: {
+                       TANACTION.TANTOKEN_COUNT: {
+                           "type": "int",
+                           "desc": _("The number of OTP values, which are "
+                                     "printed on the paper.")
+                       }
+                   }
+               }
+               }
+
+        if key:
+            ret = res.get(key, {})
+        else:
+            if ret == 'all':
+                ret = res
+        return ret
+
+    def update(self, param, reset_failcount=True):
+        tan_dict = {}
+        if "tans" in param:
+            # init tokens with tans
+            tans = param.get("tans").split()
+            tan_dict = {k: v for k, v in enumerate(tans)}
+        else:
+            # Init token without tans, so we create tans
+            param["papertoken_count"] = param.get("tantoken_count") or DEFAULT_COUNT
+            PaperTokenClass.update(self, param, reset_failcount=reset_failcount)
+            # After this creation, the init_details contain the complete list of the TANs
+            tan_dict = self.init_details.get("otps", {})
+        for tankey, tanvalue in tan_dict.iteritems():
+            # Get a 4 byte salt from the crypto module
+            salt = geturandom(SALT_LENGTH, hex=True)
+            # Now we add all TANs to the tokeninfo of this token.
+            hashed_tan = binascii.hexlify(hash(tanvalue, salt))
+            self.add_tokeninfo("tan.tan{0!s}".format(tankey),
+                               "{0}:{1}".format(binascii.hexlify(salt),
+                                               hashed_tan))
+
+    def check_otp(self, anOtpVal, counter=None, window=None, options=None):
+        """
+        check if the given OTP value is valid for this token.
+
+        :param anOtpVal: the to be verified otpvalue
+        :type anOtpVal: string
+        :param counter: the counter state, that should be verified
+        :type counter: int
+        :param window: the counter +window, which should be checked
+        :type window: int
+        :param options: the dict, which could contain token specific info
+        :type options: dict
+        :return: the counter state or -1
+        :rtype: int
+        """
+        res = -1
+        tans = self.get_tokeninfo()
+        for tankey, tanvalue in tans.iteritems():
+            if tankey.startswith("tan.tan"):
+                salt, tan = tanvalue.split(":")
+                if tan == binascii.hexlify(hash(anOtpVal,
+                                                binascii.unhexlify(salt))):
+                    self.del_tokeninfo(tankey)
+                    return 1
+
+        return res
```

Comment:
as ``res`` is not used, we could also just ``return -1`` here

Reply:
True. I thought this would be more consistent? (Or I did not think at all)
After all, we *could* return a value>0, like the number of the TAN...